app/code/Magento/Quote/Model/Cart/AddProductsToCart.php

```diff
@@ -0,0 +1,214 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\Quote\Model\Cart;
+
+use Magento\Catalog\Api\ProductRepositoryInterface;
+use Magento\Framework\Exception\NoSuchEntityException;
+use Magento\Quote\Api\CartRepositoryInterface;
+use Magento\Quote\Api\Data\CartInterface;
+use Magento\Quote\Model\Cart\BuyRequest\BuyRequestBuilder;
+use Magento\Quote\Model\Cart\Data\AddProductsToCartOutput;
+use Magento\Quote\Model\MaskedQuoteIdToQuoteIdInterface;
+use Magento\Quote\Model\Quote;
+use Magento\Framework\Message\MessageInterface;
+
+/**
+ * @inheritdoc
+ */
+class AddProductsToCart implements AddProductsToCartInterface
+{
+    /**#@+
+     * Error message codes
+     */
+    private const ERROR_PRODUCT_NOT_FOUND = 'PRODUCT_NOT_FOUND';
+    private const ERROR_INSUFFICIENT_STOCK = 'INSUFFICIENT_STOCK';
+    private const ERROR_NOT_SALABLE = 'NOT_SALABLE';
+    private const ERROR_UNDEFINED = 'UNDEFINED';
+    /**#@-*/
+
+    /**
+     * List of error messages and codes.
```

Comment:
Totally agree. I was thinking about that but then came up with the conclusion that we will need this new class on the framework level (since we will need to reuse it in different independent modules). As long as we plan to deliver this PR as a part of 2.4.0 release, I don't want to introduce new interfaces or implementations on the framework level. So, let's leave a comment for now, and then, once this PR is delivered, I will introduce a new class and refactor the existing logic. 

Thank you.